Add in-memory caching to LocaleManager

ababic opened the pull request:
The number of locales used in any one project is typically quite small (< 100), and locale objects themselves so simple and so unlikely to change, they feel like a perfect candidate for caching. Django does this for ContentTypes, which are used heavily by Wagtail. So, I figured: why not do a similar thing with Locales? So here it is...

RealOrangeOne commented on wagtail/signal_handlers.py:
```diff
@@ -117,6 +121,9 @@ def register_signal_handlers():
     post_save.connect(reset_locales_display_names_cache, sender=Locale)
     post_delete.connect(reset_locales_display_names_cache, sender=Locale)
 
+    post_delete.connect(clear_localemanager_cache, sender=Locale)
+    post_save.connect(clear_localemanager_cache, sender=Locale)
```
Issue: I'm afraid this cache clearing won't work quite as expected. Signals are only run in the thread that's handling the request, which means the cache isn't cleared in those threads.

The solution to this is to either use Redis as the cache, or architect the cache such that it doesn't need clearing (which may not be possible given data can change at any time).

ababic replied:
I think the first step here is to figure out how big a problem it is when a Locale has been deleted, but continues to be supplied, or a Locale has been updated, but the old version continues to be supplied:

- I don't think lookups by ID are too much of a problem, as `locale_id` values on objects retrieved from the database would not be referencing deleted objects
- Items cached by `language_code` are a bigger problem, but I don't think we need to worry about the 'default' Locale ever changing (it's created automatically via a migration, and changing it would cause much bigger problems). That leaves us with `get_active()` and custom `get_for_language_code()` cases... which ARE worth some more thought.

ababic replied:
I suppose the biggest risk is that someone attempts to save an object with a reference to a locale that no longer exists in the database, but I'm sure we can find a way to help prevent that.

ababic replied:
And maybe we can warn people off changing the `language_code` of an existing `Locale`, or perhaps even prevent it completely? That sounds like reasonable behaviour to me.

ababic commented:
Here are the things I plan to do to make this safer:
- Add a `WAGTAIL_LOCALEMANAGER_MAX_CACHE_AGE` setting to control how long any running instance of the app should trust data from local memory.
- Instead of caching items one-by-one, write a `populate_cache()` method to fully populate the cache, and also store a timestamp of when it was last updated
- When a locale is requested, the timestamp will be checked against `WAGTAIL_LOCALEMANAGER_MAX_CACHE_AGE`, and `populate_cache()` called to initiate a refresh before continuing.
- If no value is set for `WAGTAIL_LOCALEMANAGER_MAX_CACHE_AGE` or the value is `0`, cache validity checking, repopulation or lookups will be bypassed completely, and all queries will hit the database (making this whole thing opt-in).
- If `WAGTAIL_LOCALEMANAGER_MAX_CACHE_AGE` is set to something other than `0`, a warning will appear on the Locale edit page, mentioning that editing or deleting the locale could cause issues, and to speak to a webmaster about disabling caching temporarily (which will result in at least one app restart - clearing the cache in the process)
- A `WAGTAIL_LOCALEMANAGER_HIDE_CACHE_WARNING` setting will allow the warning to be hidden in environments where it is not relevant (e.g. in local development, or development environments where only basic single-instance hosting is used)

gasman commented:
Sounds like a lot of over-engineering to me... can we not use Django's existing cache framework here? https://docs.djangoproject.com/en/4.1/topics/cache/

ababic commented:
@gasman I personally don't think that switching from a database connection to a cache service connection (in 99% of cases) is a worthwhile swap - not when taking into account all of the change detection and repopulation that needs to happen around that. In other words, I feel that this is really only worth persuing if we use local memory.

ababic commented:
> Sounds like a lot of over-engineering to me

I would argue that the addition `referesh_cache()` simplifies things if anything, and the introduction of `WAGTAIL_LOCALEMANAGER_MAX_CACHE_AGE` feels like pretty standard practice - even outside the context of supporting multiple running instances.

gasman commented:
@ababic OK. I'm strongly against replacing something simple that works consistently (albeit not as quickly as we'd like) with something where the developer has to get their head around new obscure configuration options to decide how much they want to trade off speed for correctness.

ababic commented:
@gasman Okay, that's fair enough. What if we did this without the settings, and just caveated the docs for the following to outline that return values are fine for things like rendering a label for the current/target language, but in most other cases, a standard `get()` is the way to go?
- `LocaleManager.get_for_id()`
- `LocaleManager.get_for_language_code()`
- `TranslatableMixin.locale_obj`

ababic commented:
@gasman / @RealOrangeOne I have scaled this back somewhat, so that `get_for_language()` works exactly as it did before (where most issues would be likely to occur), and caching only applies at all to the new `get_for_id()` method.

I have also renamed `locale_obj` to `cached_obj` for improved clarity and updated documentation to explain RE the possibility of stale data.

zerolab commented:
We discussed it would be useful to get some comparison numbers.

ababic commented:
Abandoning this due to lack of interest. Feeling like threadlocals might be a decent alternative (following in the footsteps of #10018), so may resubmit again at a later time.
